core/types/celo_additions.go

```diff
@@ -1,13 +1,21 @@
 package types
 
 func (h *Header) IstanbulExtra() (*IstanbulExtra, error) {
-	if h.deserializedExtra == nil {
-		extra, err := ExtractIstanbulExtra(h)
+	h.extraLock.RLock()
+	extraData := h.deserializedExtra
+	h.extraLock.RUnlock()
+
+	if extraData == nil {
+		extra, err := extractIstanbulExtra(h)
 		if err != nil {
 			return nil, err
 		}
+		h.extraLock.Lock()
 		h.deserializedExtra = extra
+		h.extraLock.Unlock()
 	}
 
+	h.extraLock.RLock()
+	defer h.extraLock.RUnlock()
 	return h.deserializedExtra, nil
```

Comment:
I think just one lock over the whole func here would be clearer, as it stands this code could still allow the deserialization to happen multiple times.

Reply:
True, the idea was that there could be multiple readers later, but then I'll change it to use a `Mutex` instead.

Reply:
Yeah, I thought about having multiple readers too, but I don't think this is something we are going to see a lot of contention on.